src/mscorlib/src/System/Collections/Generic/Dictionary.cs

```diff
@@ -628,6 +628,49 @@ public bool Remove(TKey key)
             return false;
         }
 
+        public bool Remove(TKey key, out TValue value)
+        {
+            if (key == null)
+            {
+                value = default(TValue);
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.key);
+            }
+
+            if (buckets != null)
```

Comment:
We don't need to duplicate this for loop code, we have a private method called `FindEntry` which returns the index to that entry and if it is not found it returns -1 but it won't update the last value so we could probably add a `FindEntryRemove(key)` where it finds the entry and does the bucket/entries update and then return the value... something like:

```cs
private int FindEntryRemove(TKey key)
{
       if (buckets != null)
       {
           int hashCode = comparer.GetHashCode(key) & 0x7FFFFFFF;
           int bucket = hashCode % buckets.Length;
           int last = -1;
           for (int i = buckets[bucket]; i >= 0; last = i, i = entries[i].next)
           {
               if (entries[i].hashCode == hashCode && comparer.Equals(entries[i].key, key))
               {
                    if (last < 0)
                   {
                        buckets[bucket] = entries[i].next;
                   }
                   else
                   {
                        entries[last].next = entries[i].next;
                   }
                   return i;
               }
            }
       }
       return -1;
}
```

And then both removes could be implemented around that private API, something like:

```cs
int i = FindEntryRemove(key);
if (i > 0)
{
    value = entries[i].value;
    FreeEntry(i);
    return true;
}
value = default(TValue);
return false;
```

And we could add a private method called FreeEntry(int index), that basically does:

```cs
entries[i].hashCode = -1;
entries[i].next = freeList;
entries[i].key = default(TKey);
entries[i].value = default(TValue);
freeList = i;
freeCount++;
version++;
```

@jkotas @danmosemsft any thoughts?